File: python/115.distinct-subsequences.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        # 1: we use only the letters that met in the target
        tset = set(t)
        s = "".join(ch for ch in s if ch in tset)

        m, n = len(s), len(t)

        # 2: if given source is shorter than the target
        if m < n:
            return 0

        @cache
        def dfs(i: int, j: int) -> int:
            if j == n:
                return 1

            if i == m:
                return 0

            # 3: there is not enough symbols left to build the target
            if m - i < n - j:
                return 0

            res = 0

            # take (only if match)
            if s[i] == t[j]:
                res += dfs(i + 1, j + 1)

            # skip
            res += dfs(i + 1, j)

            return res

        # return dfs(0, 0)

        dp = [0] * n + [1] # last base case
        for i in range(m - 1, -1, -1):
            start = max(0, i - m + n)
            end = min(n, i + 1)
            for j in range(start, end):
                if s[i] == t[j]:
                    dp[j] += dp[j + 1]

        return dp[0]

File: golang/115.distinct-subsequences.go
import (
  "strings"
  "fmt"
)

func numDistinct(s string, t string) int {
    // rebuild the source string s
    // with bytes than is only contained in t  
    // we don't need other characters
    var sb strings.Builder

    tt := make(map[byte]struct{})
    for j := range len(t) {
      tt[t[j]] = struct{}{}
    }

    for i := range len(s) {
      if _, ok := tt[s[i]]; ok {
        sb.WriteByte(s[i])
      }
    }

    // the new string
    ss := sb.String()

    m, n := len(ss), len(t)
    diff := m - n


    // if the source string length (less characters)
    // than the target, therefore it's impossible to
    // find any subsequence
    if diff < 0 {
      return 0
    }

    dp := make([]int, n + 1)
    dp[n] = 1 // base case

    for i := m - 1; i >= 0; i-- {
      start := max(0, i - diff) // bottom clipping 0
      end := min(n, i + 1) // top clipping n

      for j := start; j < end; j++ {
        if ss[i] == t[j] {
          dp[j] += dp[j + 1]
        }
      }
    }

    return dp[0]
}